README.md
# topojson

A Topojson parser.

```dart
final resp = await http.get(
    Uri.parse('https://geoshape.ex.nii.ac.jp/jma/resource/AreaInformationCity_landslide/20210518/4220201.topojson'));

final TopoJson.fromJson(jsonDecode(utf8.decode(resp.bodyBytes)));
```


CHANGELOG.md
## 0.3.0

- Completely rewritten natively in Dart. No compatibility to the older versions.

## 0.2.0

- Add toGeoJson function.

## 0.1.0

- Initial version.


test/topojson_test.dart
import 'dart:convert';
import 'dart:io' as io;

import 'package:http/http.dart' as http;
import 'package:test/test.dart';
import 'package:topojson/topojson.dart';

void main() {
  group('A group of tests', () {
    setUp(() => Future.value(0));

    test('Conversion test', () async {
      final resp = await http.get(Uri.parse(
          'https://geoshape.ex.nii.ac.jp/jma/resource/AreaInformationCity_landslide/20210518/4220201.topojson'));
      final topoJson = jsonDecode(utf8.decode(resp.bodyBytes));

      TopoJson.fromJson(topoJson);

      expect(Future.value(0), completes);
    });
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:topojson_test/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/topojson.dart
enum TopoJsonObjectType {
  topology,
  point,
  multiPoint,
  lineString,
  multiLineString,
  polygon,
  multiPolygon,
  geometryCollection,
}

abstract class TopoJsonObject {
  TopoJsonObjectType get type;
  Map<String, dynamic>? get properties;

  const TopoJsonObject();

  static Map<String, dynamic>? _propFromJson(dynamic json) {
    final props = json['properties'];
    return props != null ? Map.castFrom(props) : null;
  }

  static TopoJsonObject fromJson(dynamic json, {required TopoJson topology}) {
    final String type = json['type'];
    switch (type) {
      case 'Topology':
        return TopoJson.fromJson(json);
      case 'Point':
        return TopoJsonPoint.fromJson(json);
      case 'MultiPoint':
        return TopoJsonMultiPoint.fromJson(json);
      case 'LineString':
        return TopoJsonLineString.fromJson(json, topology: topology);
      case 'MultiLineString':
        return TopoJsonMultiLineString.fromJson(json, topology: topology);
      case 'Polygon':
        return TopoJsonPolygon.fromJson(json, topology: topology);
      case 'MultiPolygon':
        return TopoJsonMultiPolygon.fromJson(json, topology: topology);
      case 'GeometryCollection':
        return TopoJsonGeometryCollection.fromJson(json, topology: topology);
      default:
        throw FormatException('Unknown object type: $type');
    }
  }

  Iterable<TopoJsonObject> iterateChildObjects();

  /// [determineObject] determines whether the iteration does continue or not by returning one of [GoSkipStopDecision] values.
  Iterable<TopoJsonObject> visitAllObjects({
    GoSkipStopDecision Function(TopoJsonObject object)? determineObject,
    bool recursive = true,
  }) =>
      _visitAllObjects(_GoSkipStopDeterminator(determineObject ?? (object) => GoSkipStopDecision.go),
          recursive: recursive);

  Iterable<TopoJsonObject> _visitAllObjects(
    _GoSkipStopDeterminator visitor, {
    bool recursive = true,
  }) sync* {
    for (final obj in iterateChildObjects()) {
      if (visitor.stop) break;
      final result = visitor.determineObject(obj);
      if (result == GoSkipStopDecision.stop) {
        visitor.stop = true;
        break;
      }
      if (result == GoSkipStopDecision.skip) continue;
      yield obj;
      if (result == GoSkipStopDecision.skipChildren || !recursive) continue;
      yield* obj._visitAllObjects(visitor, recursive: recursive);
    }
  }
}

class _GoSkipStopDeterminator {
  bool stop = false;
  final GoSkipStopDecision Function(TopoJsonObject object) determineObject;
  _GoSkipStopDeterminator(this.determineObject);
}

/// Determine iteration continuation.
enum GoSkipStopDecision {
  /// Continue iteration.
  go,

  /// Skip the object and it's children
  skip,

  /// Skip it's children. The object itself is processed (iterated).
  skipChildren,

  /// Stop the iteration process now.
  stop,
}

class TopoJson extends TopoJsonObject {
  @override
  TopoJsonObjectType get type => TopoJsonObjectType.topology;
  @override
  final Map<String, dynamic>? properties;

  final Map<String, TopoJsonObject> objects;
  final List arcs;
  final Transform transform;
  final List<num> bbox;

  const TopoJson({
    required this.objects,
    required this.arcs,
    required this.transform,
    required this.bbox,
    this.properties,
  });

  factory TopoJson.fromJson(dynamic json) {
    if (json['type'] != 'Topology') throw const FormatException('TopoJSON should have "Topology" type.');
    final tj = TopoJson(
      objects: <String, TopoJsonObject>{},
      arcs: _val(json, 'arcs'),
      transform: _transform(json),
      bbox: _val<List>(json, 'bbox').cast<num>(),
      properties: TopoJsonObject._propFromJson(json),
    );

    tj.objects.addEntries((json['objects'] as Map)
        .entries
        .map((kv) => MapEntry(kv.key, TopoJsonObject.fromJson(kv.value, topology: tj))));

    return tj;
  }

  @override
  Iterable<TopoJsonObject> iterateChildObjects() => objects.values;

  Iterable<List<num>> _getCoordsFromArcs(List arcs) {
    Iterable<List<num>>? it;
    for (final e in arcs) {
      if (it == null) {
        it = _getCoordsFromIndex(e);
      } else {
        it = it.followedBy(_getCoordsFromIndex(e).skip(1));
      }
    }
    return it ?? [];
  }

  /// Enumerate all the coordinates defined by `arcs`.
  Iterable<List<num>> _getCoordsFromIndex(int index) sync* {
    final isReversed = index < 0;
    final data = List.castFrom<dynamic, List>(arcs[isReversed ? ~index : index]);

    if (isReversed) {
      // First coord is the total of all the arcs entries
      final coord = data.fold<List<num>>(List.generate(data[0].length, (index) => 0, growable: false), _add);
      yield transform(coord);
      for (int i = data.length - 1; i > 0; i--) {
        yield transform(_subtract(coord, data[i]));
      }
    } else {
      final coord = data[0].cast<num>();
      yield transform(coord);
      for (int i = 1; i < data.length; i++) {
        yield transform(_add(coord, data[i]));
      }
    }
  }

  static Transform _transform(dynamic json) {
    final t = json['transform'];
    if (t == null) {
      return (coords) => coords; // identity transform
    }
    final scale = _val<List>(t, 'scale', context: 'transform').cast<num>();
    final translate = _val<List>(t, 'translate', context: 'transform').cast<num>();
    return (coords) =>
        List.generate(coords.length, (index) => coords[index] * scale[index] + translate[index], growable: false);
  }

  static T _val<T>(dynamic json, String name, {String? context}) {
    final v = json[name];
    if (v is T) return v;
    if (context != null) {
      throw FormatException('$context does not have field named "$name" of $T.');
    }
    throw FormatException('No field named "$name" of $T.');
  }

  static List<num> _add(List<num> accum, List delta) {
    final d = delta.cast<num>();
    for (int i = 0; i < accum.length; i++) {
      accum[i] += d[i];
    }
    return accum;
  }

  static List<num> _subtract(List<num> accum, List delta) {
    final d = delta.cast<num>();
    for (int i = 0; i < accum.length; i++) {
      accum[i] -= d[i];
    }
    return accum;
  }
}

typedef Transform = List<num> Function(List<num> coords);

class TopoJsonGeometryCollection extends TopoJsonObject {
  @override
  TopoJsonObjectType get type => TopoJsonObjectType.geometryCollection;
  @override
  final Map<String, dynamic>? properties;

  final List<TopoJsonObject> geometries;

  TopoJsonGeometryCollection({
    required this.geometries,
    required this.properties,
  });

  factory TopoJsonGeometryCollection.fromJson(
    dynamic json, {
    required TopoJson topology,
  }) =>
      TopoJsonGeometryCollection(
        geometries: (json['geometries'] as List)
            .map(
              (g) => TopoJsonObject.fromJson(
                g,
                topology: topology,
              ),
            )
            .toList(),
        properties: TopoJsonObject._propFromJson(json),
      );

  @override
  Iterable<TopoJsonObject> iterateChildObjects() => geometries;
}

class TopoJsonPoint extends TopoJsonObject {
  @override
  TopoJsonObjectType get type => TopoJsonObjectType.point;
  @override
  final Map<String, dynamic>? properties;

  final List<num> coordinates;

  TopoJsonPoint({
    required this.coordinates,
    this.properties,
  });

  factory TopoJsonPoint.fromJson(dynamic json) {
    return TopoJsonPoint(
      coordinates: ((json['coordinates'] as List).first as List).cast<num>(),
      properties: json['properties'],
    );
  }

  /// Do nothing for [TopoJsonPoint].
  @override
  Iterable<TopoJsonObject> iterateChildObjects() => const [];
}

class TopoJsonMultiPoint extends TopoJsonObject {
  @override
  TopoJsonObjectType get type => TopoJsonObjectType.multiPoint;
  @override
  final Map<String, dynamic>? properties;

  final List<TopoJsonPoint> points;

  TopoJsonMultiPoint({
    required this.points,
    this.properties,
  });

  factory TopoJsonMultiPoint.fromJson(dynamic json) {
    return TopoJsonMultiPoint(
      points: (json['coordinates'] as List).map((c) => TopoJsonPoint(coordinates: (c as List).cast<num>())).toList(),
      properties: json['properties'],
    );
  }

  @override
  Iterable<TopoJsonObject> iterateChildObjects() => points;
}

class TopoJsonLineString extends TopoJsonObject {
  @override
  TopoJsonObjectType get type => TopoJsonObjectType.lineString;
  @override
  final Map<String, dynamic>? properties;

  final List<List<num>> points;

  TopoJsonLineString({
    required this.points,
    this.properties,
  });

  factory TopoJsonLineString.fromJson(dynamic json, {required TopoJson topology}) {
    return TopoJsonLineString(
      points: topology._getCoordsFromArcs(json['arcs']).toList(),
      properties: json['properties'],
    );
  }

  /// Do noting for [TopoJsonLineString].
  @override
  Iterable<TopoJsonObject> iterateChildObjects() => const [];
}

class TopoJsonMultiLineString extends TopoJsonObject {
  @override
  TopoJsonObjectType get type => TopoJsonObjectType.multiLineString;
  @override
  final Map<String, dynamic>? properties;

  final List<TopoJsonLineString> lineStrings;

  TopoJsonMultiLineString({
    required this.lineStrings,
    this.properties,
  });

  factory TopoJsonMultiLineString.fromJson(dynamic json, {required TopoJson topology}) {
    return TopoJsonMultiLineString.fromArcs(
      json['arcs'],
      topology: topology,
      properties: json['properties'],
    );
  }

  factory TopoJsonMultiLineString.fromArcs(List arcs, {required TopoJson topology, Map<String, dynamic>? properties}) {
    return TopoJsonMultiLineString(
      lineStrings: arcs.map((a) => TopoJsonLineString(points: topology._getCoordsFromArcs(a).toList())).toList(),
      properties: properties,
    );
  }

  @override
  Iterable<TopoJsonObject> iterateChildObjects() => lineStrings;
}

class TopoJsonRing {
  final List<List<num>> points;
  const TopoJsonRing({required this.points});
  factory TopoJsonRing.fromJson(dynamic json, {required TopoJson topology}) {
    return TopoJsonRing(
      points: topology._getCoordsFromArcs(json['arcs']).toList(),
    );
  }
}

class TopoJsonPolygon extends TopoJsonObject {
  @override
  TopoJsonObjectType get type => TopoJsonObjectType.polygon;
  @override
  final Map<String, dynamic>? properties;

  final List<TopoJsonRing> rings;

  TopoJsonPolygon({
    required this.rings,
    this.properties,
  });

  factory TopoJsonPolygon.fromJson(dynamic json, {required TopoJson topology}) {
    return TopoJsonPolygon.fromArcs(
      json['arcs'],
      topology: topology,
      properties: json['properties'],
    );
  }

  factory TopoJsonPolygon.fromArcs(List arcs, {required TopoJson topology, Map<String, dynamic>? properties}) {
    return TopoJsonPolygon(
      rings: arcs.map((a) => TopoJsonRing(points: topology._getCoordsFromArcs(a).toList())).toList(),
      properties: properties,
    );
  }

  /// Do nothing for [TopoJsonPolygon].
  @override
  Iterable<TopoJsonObject> iterateChildObjects() => const [];
}

class TopoJsonMultiPolygon extends TopoJsonObject {
  @override
  TopoJsonObjectType get type => TopoJsonObjectType.multiPolygon;
  @override
  final Map<String, dynamic>? properties;

  final List<TopoJsonPolygon> polygons;

  TopoJsonMultiPolygon({
    required this.polygons,
    this.properties,
  });

  factory TopoJsonMultiPolygon.fromJson(dynamic json, {required TopoJson topology}) {
    return TopoJsonMultiPolygon(
      polygons: (json['arcs'] as List).map((a) => TopoJsonPolygon.fromArcs(a, topology: topology)).toList(),
      properties: json['properties'],
    );
  }

  @override
  Iterable<TopoJsonObject> iterateChildObjects() => polygons;
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"topojson_test", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
import 'dart:convert';
import 'dart:ui';

import 'package:http/http.dart' as http;
import 'package:flutter/material.dart';

import 'package:topojson/topojson.dart';

void main() async {
  final resp = await http.get(Uri.parse(
      'https://geoshape.ex.nii.ac.jp/jma/resource/AreaInformationCity_landslide/20210518/4220201.topojson'));
  final json = jsonDecode(utf8.decode(resp.bodyBytes));
  final tj = TopoJson.fromJson(json);

  runApp(MyApp(
    topojson: tj,
  ));
}

class MyApp extends StatelessWidget {
  final TopoJson topojson;
  const MyApp({Key? key, required this.topojson}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    final bbox = topojson.bbox.map((n) => n.toDouble()).toList();
    final rect = Rect.fromLTRB(bbox[0], bbox[1], bbox[2], bbox[3]);
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        body: LayoutBuilder(builder: (context, constraints) {
          return Container(
            width: constraints.maxWidth,
            height: constraints.maxHeight,
            //color: Colors.blue,
            child: CustomPaint(
              size: Size(constraints.maxWidth, constraints.maxHeight),
              painter: TopoJsonPainter(topojson, rect),
            ),
          );
        }),
      ),
    );
  }
}

class TopoJsonPainter extends CustomPainter {
  final TopoJson topojson;
  final Rect rect;
  const TopoJsonPainter(this.topojson, this.rect);

  /// NOTE: This sample draws the geometry data without any coordinate conversion/correction; testing purpose only.
  @override
  void paint(Canvas canvas, Size size) {
    for (final object in topojson.visitAllObjects()) {
      if (object.type == TopoJsonObjectType.polygon) {
        final paint = Paint()
          ..color = Colors.redAccent
          ..style = PaintingStyle.stroke
          ..strokeWidth = 0.5;
        // Draw as if lat/lng values were in cartesian coordinate system
        for (final ring in (object as TopoJsonPolygon).rings) {
          canvas.drawPoints(
              PointMode.polygon,
              ring.points
                  .map((p) => Offset(
                        (p[0].toDouble() - rect.left) / rect.width * size.width,
                        (p[1].toDouble() - rect.top) /
                            rect.height *
                            size.height,
                      ))
                  .toList(),
              paint);
        }
      }
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}
